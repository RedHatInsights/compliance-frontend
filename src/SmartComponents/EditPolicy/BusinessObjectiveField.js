import React from 'react';
import { compose } from 'redux';
import { Field, reduxForm } from 'redux-form';
import { FormGroup, Title } from '@patternfly/react-core';
import propTypes from 'prop-types';
import { ReduxFormCreatableSelectInput } from 'PresentationalComponents/ReduxFormWrappers/ReduxFormWrappers';
import gql from 'graphql-tag';
import { withApollo } from 'react-apollo';

const GET_BUSINESS_OBJECTIVES = gql`
query businessObjectives {
    businessObjectives {
        id
        title
    }
}
`;

class BusinessObjectiveField extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            selected: props.businessObjective ? this.createOption(props.businessObjective) : null
        };
    }

    loadOptions = (title) => {
        const variables = title ? { title } : {};
        this.props.client.cache.reset();
        return this.props.client.query({
            query: GET_BUSINESS_OBJECTIVES,
            variables
        }).then((items) => {
            return items.data.businessObjectives.map(businessObjective => (
                this.createOption(businessObjective)
            ));
        });
    }

    createOption = (businessObjective) => ({
        label: businessObjective.title,
        value: businessObjective.id
    });

    handleChange = (newValue) => {
        const { dispatch } = this.props;
        this.setState({ selected: newValue });

        // Manually dispatch the action to ensure the newly created label is set
        dispatch({
            type: '@@redux-form/CHANGE',
            meta: {
                field: 'businessObjective',
                form: 'editPolicy'
            },
            payload: newValue
        });
    };

    handleCreate = (inputValue) => {
        let newOption = this.createOption({ title: inputValue, value: inputValue });
        newOption.create = true;
        this.handleChange(newOption);
    };

    render() {
        const { selected } = this.state;
        const titleId = 'business-objective-typeahead';

        return (
            <React.Fragment>
                <Title headingLevel='h3' size='xl'>Business objective</Title>
                This is an optional field that can be used to label policies that are related to
                specific business objectives.
                <FormGroup field-id='edit-policy-business-objective'
                    label="Business objective"
                    helperText='e.g Project Gemini'>
                    <Field name='businessObjective'
                        id='businessObjective'
                        ariaLabelledBy={titleId}
                        aria-label="Select a business objective"
                        component={ReduxFormCreatableSelectInput}
                        isClearable
                        placeholder='Type to select and create'
                        onChange={this.handleChange}
                        onCreateOption={this.handleCreate}
                        loadOptions={this.loadOptions}
                        selected={selected}
                    />
                </FormGroup>
            </React.Fragment>
        );
    }
}

BusinessObjectiveField.propTypes = {
    businessObjective: propTypes.object,
    client: propTypes.object,
    dispatch: propTypes.func
};

export default compose(
    withApollo,
    reduxForm({
        form: 'editPolicy'
    })
)(BusinessObjectiveField);
