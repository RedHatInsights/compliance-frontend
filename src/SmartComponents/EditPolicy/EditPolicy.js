import {
    Modal,
    Form,
    Button
} from '@patternfly/react-core';
import React, { Component } from 'react';
import propTypes from 'prop-types';
import UpdateProfileButton from './UpdateProfileButton';
import {
    ProfileThresholdField,
    BusinessObjectiveField
} from 'SmartComponents';
import { formValueSelector } from 'redux-form';
import { connect } from 'react-redux';
import round from 'lodash/round';

export class EditPolicy extends Component {
    constructor(props) {
        super(props);
        this.state = {
            isOpen: false,
            isModalOpen: false,
            policyId: props.policyId,
            businessObjective: props.businessObjective
        };
    }

    handleModalToggle = () => {
        const { isModalOpen } = this.state;
        insights.chrome.appAction(!isModalOpen ? 'edit-business-objective' : '');
        this.setState(() => ({
            isModalOpen: !isModalOpen,
            businessObjective: this.props.businessObjective
        }));
        this.props.onClose();
    };

    render() {
        const { policyId, isModalOpen, businessObjective } = this.state;
        const { previousThreshold, editPolicyBusinessObjective, complianceThreshold, dispatch } = this.props;

        return (
            <React.Fragment>
                <Button
                    variant='secondary'
                    onClick={this.handleModalToggle}
                >
                    Edit policy
                </Button>
                <Modal
                    isSmall
                    title="Edit policy details"
                    isOpen={isModalOpen}
                    onClose={this.handleModalToggle}
                    isFooterLeftAligned
                    actions={[
                        <Button key="cancel" variant="secondary" onClick={this.handleModalToggle}>
                            Cancel
                        </Button>,
                        <UpdateProfileButton
                            key='confirm'
                            policyId={policyId}
                            threshold={ round(parseFloat(complianceThreshold || previousThreshold), 1) }
                            businessObjective={businessObjective}
                            editPolicyBusinessObjective={editPolicyBusinessObjective}
                            onClick={this.handleModalToggle}
                        />
                    ]}
                >
                    <Form>
                        <BusinessObjectiveField
                            businessObjective={businessObjective}
                            policyId={policyId}
                            dispatch={dispatch}
                        />
                        <ProfileThresholdField previousThreshold={previousThreshold} />
                    </Form>
                </Modal>
            </React.Fragment>
        );
    }
}

EditPolicy.propTypes = {
    policyId: propTypes.string,
    previousThreshold: propTypes.number,
    businessObjective: propTypes.object,
    editPolicyBusinessObjective: propTypes.object,
    complianceThreshold: propTypes.string,
    onClose: propTypes.func,
    dispatch: propTypes.func
};

const selector = formValueSelector('policyForm');
export default connect(
    state => ({
        complianceThreshold: selector(state, 'complianceThreshold'),
        editPolicyBusinessObjective: selector(state, 'businessObjective')
    })
)(EditPolicy);
