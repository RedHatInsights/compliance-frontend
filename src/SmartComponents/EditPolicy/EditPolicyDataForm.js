import React from 'react';
import propTypes from 'prop-types';
import { FormRenderer } from 'PresentationalComponents';
import { mapCountOsMinorVersions } from 'Store/Reducers/SystemStore';
import { default as formSchema, componentMapper } from './schema';

const profilesToOsMinorMap = (profiles, hosts) => (
    (profiles || []).reduce((acc, profile) => {
        if (profile.osMinorVersion !== '') {
            acc[profile.osMinorVersion] ||= { osMinorVersion: profile.osMinorVersion, count: 0 };
        }

        return acc;
    }, mapCountOsMinorVersions(hosts || []))
);

export const EditPolicyForm = ({ policy }) => {
    const policyProfiles = policy?.policy?.profiles || [];
    const osMinorVersionCounts = profilesToOsMinorMap(policyProfiles, policy.hosts);

    // Using JSON in a string datatype may not be ideal
    // look into either being able to define a custom data type or solve with other more fitting type
    const initialRuleSelection = JSON.stringify(policyProfiles.map((policyProfile) => ({
        id: policyProfile.id,
        ruleRefIds: policyProfile.rules.map((rule) => (rule.refId)),
        count: osMinorVersionCounts[policyProfile.osMinorVersion]
    })));
    const initialSystemsSeletion = JSON.stringify(policy?.hosts);

    const schema = formSchema(policy, {
        initialRuleSelection,
        initialSystemsSeletion,
        ruleFieldProps: {
            osMinorVersionCounts,
            policy
        }
    });

    const onSubmit = (...args) => {
        console.log(...args);
    };

    const onCancel = (...args) => {
        console.log(...args);
    };

    return <FormRenderer { ...{
        componentMapper,
        schema,
        onSubmit,
        onCancel
    } } />;
};

EditPolicyForm.propTypes = {
    policy: propTypes.object,
    updatedPolicy: propTypes.object,
    setUpdatedPolicy: propTypes.func
};

export default EditPolicyForm;
