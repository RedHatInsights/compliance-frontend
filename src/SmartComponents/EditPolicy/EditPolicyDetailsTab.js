import React, { useState } from 'react';
import propTypes from 'prop-types';
import {
    FormGroup, TextArea, TextInput
} from '@patternfly/react-core';
import {
    PolicyThresholdTooltip,
    PolicyBusinessObjectiveTooltip
} from 'PresentationalComponents';
import { thresholdValid } from '../CreatePolicy/validate';

export const useThresholdValidate = () => {
    const [validThreshold, setValidThreshold] = useState(true);
    return [validThreshold, (threshold) => {
        const valid = thresholdValid(threshold);
        setValidThreshold(valid);
        return valid;
    }];
};

const EditPolicyDetailsTab = ({ policy, setUpdatedPolicy }) => {
    const [validThreshold, validateThreshold] = useThresholdValidate();

    return <div className="pf-c-form">
        <FormGroup label="Policy description" isRequired fieldId="description">
            <TextArea
                style={ { width: 800, height: 110 } }
                isRequired
                type="text"
                id="description"
                name="description"
                aria-describedby="description"
                defaultValue={ policy.description }
                onChange={ (value) => {
                    setUpdatedPolicy((policy) => ({
                        ...policy,
                        description: value
                    }));
                }} />
        </FormGroup>

        <FormGroup
            label="Business objective"
            labelIcon={ <PolicyBusinessObjectiveTooltip /> }
            fieldId="business-objective">
            <TextInput
                type="text"
                style={ { width: 300 } }
                id="business-objective"
                name="business-objective"
                aria-describedby="business-objective"
                defaultValue={ policy?.businessObjective?.title }
                onChange={ (value) => {
                    setUpdatedPolicy((policy) =>({
                        ...policy,
                        businessObjective: {
                            ...policy.businessObjective,
                            title: value
                        }
                    }));
                }} />
        </FormGroup>

        <FormGroup
            validated={ validThreshold ? 'default' : 'error' }
            label='Compliance threshold (%)'
            labelIcon={ <PolicyThresholdTooltip /> }
            fieldId='policy-threshold'
            helperTextInvalid='Threshold has to be a number between 0 and 100'
            helperText='A value of 95% or higher is recommended'>
            <TextInput
                type='number'
                style={ { width: 150 } }
                name='compliance-threshold'
                id='compliance-threshold'
                defaultValue={ policy.complianceThreshold }
                aria-describedby="policy-threshold"
                onChange={ (value) => {
                    setUpdatedPolicy((policy) =>({
                        ...policy,
                        complianceThreshold: value,
                        complianceThresholdValid: validateThreshold(value)
                    }));
                }} />
        </FormGroup>
    </div>;
};

EditPolicyDetailsTab.propTypes = {
    policy: propTypes.shape({
        description: propTypes.string,
        businessObjective: propTypes.object,
        complianceThreshold: propTypes.oneOfType([
            propTypes.string,
            propTypes.number
        ])
    }),
    setUpdatedPolicy: propTypes.func
};

export default EditPolicyDetailsTab;
