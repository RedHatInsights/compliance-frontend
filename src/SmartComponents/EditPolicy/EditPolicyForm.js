import React, { useCallback, useState, useEffect } from 'react';
import propTypes from 'prop-types';
import { Form, Tab, TabTitleText } from '@patternfly/react-core';
import { RoutedTabs } from 'PresentationalComponents';
import EditPolicyRulesTab from './EditPolicyRulesTab';
import EditPolicySystemsTab from './EditPolicySystemsTab';
import NewRulesAlert from './components/NewRulesAlert';
import { mapCountOsMinorVersions } from 'Store/Reducers/SystemStore';
import { profilesWithRulesToSelection } from 'PresentationalComponents/TabbedRules';
import { thresholdValid } from '../CreatePolicy/validate';
import { useNewRulesAlertState } from './hooks/index';

const profilesToOsMinorMap = (profiles, hosts) =>
  (profiles || []).reduce((acc, profile) => {
    if (profile.osMinorVersion !== '') {
      acc[profile.osMinorVersion] ||= {
        osMinorVersion: profile.osMinorVersion,
        count: 0,
      };
    }

    return acc;
  }, mapCountOsMinorVersions(hosts || []));

const EditPolicyForm = ({
  policy,
  setUpdatedPolicy,
  selectedRuleRefIds,
  setSelectedRuleRefIds,
  selectedSystems,
  setSelectedSystems,
  setRuleValues,
  ruleValues,
}) => {
  const policyProfiles = policy?.policy?.profiles || [];
  const [osMinorVersionCounts, setOsMinorVersionCounts] = useState({});
  const [newRulesAlert, setNewRulesAlert] = useNewRulesAlertState(false);

  const handleSystemSelect = useCallback(
    (newSelectedSystems) => {
      const policyMinorVersions = policy.hosts.map(
        ({ osMinorVersion }) => osMinorVersion
      );
      const hasNewOsMinorVersions =
        newSelectedSystems.filter(
          ({ osMinorVersion }) => !policyMinorVersions.includes(osMinorVersion)
        ).length > 0;

      setSelectedSystems(newSelectedSystems);
      setNewRulesAlert(hasNewOsMinorVersions);
      setOsMinorVersionCounts(
        profilesToOsMinorMap(policyProfiles, newSelectedSystems)
      );
    },
    [policyProfiles, selectedRuleRefIds]
  );

  useEffect(() => {
    if (policy) {
      const complianceThresholdValid = thresholdValid(
        policy.complianceThreshold
      );
      const profilesWithOsMinor = policyProfiles.filter(
        ({ osMinorVersion }) => !!osMinorVersion
      );
      setUpdatedPolicy({
        ...policy,
        complianceThresholdValid,
      });

      setSelectedRuleRefIds(profilesWithRulesToSelection(profilesWithOsMinor));
      handleSystemSelect(policy.hosts);
    }
  }, [policy]);

  return (
    <Form>
      <RoutedTabs ouiaId="EditSystems" defaultTab="systems" id="policy-tabs">
        <Tab
          eventKey="rules"
          ouiaId="Rules"
          title={<TabTitleText>Rules</TabTitleText>}
        >
          <EditPolicyRulesTab
            policy={policy}
            setSelectedRuleRefIds={setSelectedRuleRefIds}
            setRuleValues={setRuleValues}
            ruleValues={ruleValues}
            selectedRuleRefIds={selectedRuleRefIds}
            osMinorVersionCounts={osMinorVersionCounts}
          />
        </Tab>
        <Tab
          eventKey="systems"
          ouiaId="Systems"
          title={<TabTitleText>Systems</TabTitleText>}
        >
          <EditPolicySystemsTab
            policy={policy}
            selectedSystems={selectedSystems}
            onSystemSelect={handleSystemSelect}
          />
          {newRulesAlert && <NewRulesAlert />}
        </Tab>
      </RoutedTabs>
    </Form>
  );
};

EditPolicyForm.propTypes = {
  policy: propTypes.object,
  updatedPolicy: propTypes.object,
  setUpdatedPolicy: propTypes.func,
  selectedRuleRefIds: propTypes.arrayOf(propTypes.object),
  setSelectedRuleRefIds: propTypes.func,
  setSelectedSystems: propTypes.func,
  selectedSystems: propTypes.array,
  setRuleValues: propTypes.func,
  ruleValues: propTypes.array,
};

export default EditPolicyForm;
