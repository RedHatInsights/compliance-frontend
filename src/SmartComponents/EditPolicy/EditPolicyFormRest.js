import React, { useCallback, useState } from 'react';
import propTypes from 'prop-types';
import { Form, Tab, TabTitleText } from '@patternfly/react-core';
import { RoutedTabs } from 'PresentationalComponents';
import EditPolicyRulesTabRest from './EditPolicyRulesTabRest';
import EditPolicySystemsTabRest from './EditPolicySystemsTabRest';
import NewRulesAlert from './components/NewRulesAlert';
import { useNewRulesAlertState } from './hooks/index';

const getCounts = (arr) =>
  arr.reduce(
    (prev, cur) => ({
      ...prev,
      [cur]: prev[cur] ? prev[cur] + 1 : 1,
    }),
    {}
  );

const EditPolicyForm = ({
  policy,
  setUpdatedPolicy,
  assignedRuleIds,
  assignedSystems,
  setRuleValues,
  supportedOsVersions,
}) => {
  const [selectedOsMinorVersions, setSelectedOsMinorVersions] = useState([
    ...new Set(assignedSystems.map((system) => system.os_minor_version)),
  ]);
  const [selectedVersionCounts, setSelectedVersionCounts] = useState(
    getCounts(assignedSystems.map((system) => system.os_minor_version))
  );
  const [newRulesAlert, setNewRulesAlert] = useNewRulesAlertState(false);

  const [selectedSystems, setSelectedSystems] = useState(
    assignedSystems?.map((system) => system.id)
  );
  const preUsedOsMinorVersions = assignedSystems.map(
    (system) => system.os_minor_version
  );

  const handleSystemSelect = useCallback(
    (newSelectedSystems) => {
      const newOsMinorVersions = [
        ...new Set(newSelectedSystems.map((system) => system.osMinorVersion)), // get unique values
      ];

      const hasNewOsMinorVersions =
        newOsMinorVersions.filter(
          (osMinorVersion) => !preUsedOsMinorVersions.includes(osMinorVersion)
        ).length > 0;

      setUpdatedPolicy((prev) => ({
        ...prev,
        hosts: newSelectedSystems.map((system) => system.id),
      }));
      setNewRulesAlert(hasNewOsMinorVersions);
      setSelectedOsMinorVersions(newOsMinorVersions);
      setSelectedVersionCounts(
        getCounts(newSelectedSystems.map((system) => system.osMinorVersion))
      );
      setSelectedSystems(newSelectedSystems.map((system) => system.id));
    },
    [preUsedOsMinorVersions, setNewRulesAlert, setUpdatedPolicy]
  );

  return (
    <Form>
      <RoutedTabs ouiaId="EditSystems" defaultTab="rules" id="policy-tabs">
        <Tab
          eventKey="rules"
          ouiaId="Rules"
          title={<TabTitleText>Rules</TabTitleText>}
        >
          <EditPolicyRulesTabRest
            policy={policy}
            setRuleValues={setRuleValues}
            assignedRuleIds={assignedRuleIds}
            selectedOsMinorVersions={selectedOsMinorVersions}
            selectedVersionCounts={selectedVersionCounts}
            setUpdatedPolicy={setUpdatedPolicy}
          />
        </Tab>
        <Tab
          eventKey="systems"
          ouiaId="Systems"
          title={<TabTitleText>Systems</TabTitleText>}
        >
          <EditPolicySystemsTabRest
            policy={policy}
            selectedSystems={selectedSystems}
            onSystemSelect={handleSystemSelect}
            supportedOsVersions={supportedOsVersions}
          />
          {newRulesAlert && <NewRulesAlert />}
        </Tab>
      </RoutedTabs>
    </Form>
  );
};

EditPolicyForm.propTypes = {
  policy: propTypes.object,
  updatedPolicy: propTypes.object,
  setUpdatedPolicy: propTypes.func,
  assignedRuleIds: propTypes.arrayOf(propTypes.object),
  assignedSystems: propTypes.array,
  setRuleValues: propTypes.func,
  supportedOsVersions: propTypes.array,
};

export default EditPolicyForm;
