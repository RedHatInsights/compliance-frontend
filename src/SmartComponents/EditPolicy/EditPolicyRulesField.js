import React from 'react';
import {
    EmptyState, EmptyStateBody, Text, TextContent, Title
} from '@patternfly/react-core';
import propTypes from 'prop-types';
import useFieldApi from '@data-driven-forms/react-form-renderer/use-field-api';
import EmptyTable from '@redhat-cloud-services/frontend-components/EmptyTable';
import Spinner from '@redhat-cloud-services/frontend-components/Spinner';
import { StateViewWithError, StateViewPart, TabbedRules } from 'PresentationalComponents';
import {
    useBenchmarksQuery, useTabsData, useProfilesQuery
} from './hooks';

const EditPolicyRulesTabEmptyState = () => <EmptyState>
    <Title headingLevel="h5" size="lg">
        No rules can be configured
    </Title>
    <EmptyStateBody>
        This policy has no associated systems, and therefore no rules can be configured.
    </EmptyStateBody>
    <EmptyStateBody>
        Add at least one system to configure rules for this policy.
    </EmptyStateBody>
</EmptyState>;

export const EditPolicyRulesField = (props) => {
    const { input } = useFieldApi(props);
    const {
        FieldProps: { policy, osMinorVersionCounts }
    } = props;
    const toTabsData = useTabsData(policy, osMinorVersionCounts);
    const selectedRuleRefIds = JSON.parse(input.value);

    const {
        data: benchmarksData,
        error: benchmarksError,
        loading: benchmarksLoading
    } = useBenchmarksQuery(policy, osMinorVersionCounts);
    const benchmarks = benchmarksData?.benchmarks?.nodes;

    const tabsData = toTabsData(benchmarks, selectedRuleRefIds);

    const {
        data: profilesData, error: profilesError, loading: profilesLoading
    } = useProfilesQuery(tabsData);

    const loadingState = ((profilesLoading || benchmarksLoading) ? true : undefined);
    const dataState = ((!loadingState && tabsData?.length > 0) ? profilesData : undefined);
    const error = benchmarksError || profilesError;

    const handleSelect = ({ id: profileId }, rules) => {
        const updatedProfileRules = selectedRuleRefIds.map((ruleSet) => {
            if (ruleSet.id === profileId) {
                return { ...ruleSet, ruleRefIds: rules };
            } else {
                return ruleSet;
            }
        });

        input.onChange(JSON.stringify(updatedProfileRules));
    };

    return <StateViewWithError stateValues={ {
        error,
        data: !error && dataState,
        loading: loadingState,
        empty: !loadingState && !dataState && !error
    } }>
        <StateViewPart stateKey="loading">
            <EmptyTable><Spinner/></EmptyTable>
        </StateViewPart>
        <StateViewPart stateKey="data">
            <TextContent>
                <Text>
                    Different release versions of RHEL are associated with different versions of
                    the SCAP Security Guide (SSG), therefore each release must be customized independently.
                </Text>
            </TextContent>
            <TabbedRules
                tabsData={ tabsData }
                remediationsEnabled={ false }
                selectedFilter
                level={ 1 }
                handleSelect={ handleSelect } />
        </StateViewPart>
        <StateViewPart stateKey="empty">
            <EditPolicyRulesTabEmptyState />
        </StateViewPart>
    </StateViewWithError>;
};

EditPolicyRulesField.propTypes = {
    policy: propTypes.object,
    osMinorVersionCounts: propTypes.shape({
        osMinorVersion: propTypes.shape({
            osMinorVersion: propTypes.number,
            count: propTypes.number
        })
    }),
    FieldProps: propTypes.object
};

export default EditPolicyRulesField;
