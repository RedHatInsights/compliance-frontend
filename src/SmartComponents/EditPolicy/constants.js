import { gql } from '@apollo/client';

export const BENCHMARKS_QUERY = gql`
  query EP_Benchmarks($filter: String!) {
    benchmarks(search: $filter) {
      nodes {
        id
        latestSupportedOsMinorVersions
        ruleTree
        valueDefinitions {
          defaultValue
          description
          id
          refId
          title
          valueType
        }
        profiles {
          id
          refId
          ssgVersion
        }
      }
    }
  }
`;

export const MULTIVERSION_QUERY = gql`
  query EP_Profile($policyId: String!) {
    profile(id: $policyId) {
      id
      name
      refId
      external
      description
      totalHostCount
      compliantHostCount
      complianceThreshold
      osMajorVersion
      supportedOsVersions
      lastScanned
      policyType
      policy {
        id
        name
        refId
        profiles {
          id
          parentProfileId
          name
          refId
          osMinorVersion
          osMajorVersion
          values
          benchmark {
            id
            title
            latestSupportedOsMinorVersions
            osMajorVersion
            version
            ruleTree
          }
          rules {
            title
            severity
            rationale
            refId
            description
            remediationAvailable
            identifier
            values
          }
        }
      }
      businessObjective {
        id
        title
      }
      hosts {
        id
        osMinorVersion
        osMajorVersion
      }
    }
  }
`;

export const RULE_VALUE_DEFINITIONS_QUERY = gql`
  query EP_ProfileValueDefinitions($policyId: String!) {
    profile(id: $policyId) {
      id
      policy {
        id
        refId
        profiles {
          id
          parentProfileId
          refId
          benchmark {
            id
            ruleTree
            valueDefinitions {
              defaultValue
              description
              id
              refId
              title
              valueType
            }
          }
        }
      }
    }
  }
`;
