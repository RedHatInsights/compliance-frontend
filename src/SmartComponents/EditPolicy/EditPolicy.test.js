import toJson from 'enzyme-to-json';

import { EditPolicy } from './EditPolicy.js';

describe('EditPolicy', () => {
    const defaultProps = {
        policyId: 'POLICY_ID',
        businessObjective: {
            title: 'BO title',
            id: 1
        },
        complianceThreshold: '30',
        onClose: jest.fn(),
        dispatch: jest.fn()
    };

    it('expect to render without error', () => {
        const wrapper = shallow(
            <EditPolicy { ...defaultProps }/>
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
