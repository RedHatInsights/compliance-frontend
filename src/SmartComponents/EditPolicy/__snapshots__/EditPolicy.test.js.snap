// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditPolicy expect to render without error 1`] = `
<Fragment>
  <Dropdown
    className="policy-details-dropdown"
    dropdownItems={
      Array [
        <DropdownItem
          component="button"
          onClick={[Function]}
        >
          Edit policy
        </DropdownItem>,
      ]
    }
    isOpen={false}
    onSelect={[Function]}
    position="right"
    toggle={
      <DropdownToggle
        onToggle={[Function]}
      >
        Actions
      </DropdownToggle>
    }
  />
  <Modal
    actions={
      Array [
        <Unknown
          onClick={[Function]}
          variant="secondary"
        >
          Cancel
        </Unknown>,
        <Apollo(UpdateProfileButton)
          businessObjective={
            Object {
              "id": 1,
              "title": "BO title",
            }
          }
          onClick={[Function]}
          policyId="POLICY_ID"
          threshold={30}
        />,
      ]
    }
    appendTo={<body />}
    ariaDescribedById=""
    className=""
    hideTitle={false}
    isFooterLeftAligned={true}
    isLarge={false}
    isOpen={false}
    isSmall={true}
    onClose={[Function]}
    showClose={true}
    title="Edit policy details"
  >
    <Form>
      <withApollo(ReduxForm)
        businessObjective={
          Object {
            "id": 1,
            "title": "BO title",
          }
        }
        dispatch={[MockFunction]}
        policyId="POLICY_ID"
      />
      <ReduxForm
        destroyOnUnmount={true}
        enableReinitialize={false}
        forceUnregisterOnUnmount={false}
        form="editPolicy"
        getFormState={[Function]}
        keepDirtyOnReinitialize={false}
        persistentSubmitErrors={false}
        pure={true}
        shouldAsyncValidate={[Function]}
        shouldError={[Function]}
        shouldValidate={[Function]}
        shouldWarn={[Function]}
        submitAsSideEffect={false}
        touchOnBlur={true}
        touchOnChange={false}
        updateUnregisteredFields={false}
      />
    </Form>
  </Modal>
</Fragment>
`;
